File: packages/node/test/deployment.test.ts
import { Bicep } from '@azure/bicep-rpc-client';
import { TokenCredential } from '@azure/core-auth';
import { DeploymentStacksClient } from '@azure/arm-resourcesdeploymentstacks';
import {
  BicepTestSession,
  DeployOptions,
  LiveBicepTestSession,
} from '../src';

const createAtResourceGroup = jest.fn();
const createAtSubscription = jest.fn();
const createAtManagementGroup = jest.fn();
const validateAtResourceGroup = jest.fn();
const validateAtSubscription = jest.fn();
const validateAtManagementGroup = jest.fn();
const deleteAtResourceGroup = jest.fn();
const deleteAtSubscription = jest.fn();
const deleteAtManagementGroup = jest.fn();

jest.mock('@azure/arm-resourcesdeploymentstacks', () => ({
  DeploymentStacksClient: jest.fn().mockImplementation(() => ({
    deploymentStacks: {
      createOrUpdateAtResourceGroup: createAtResourceGroup,
      createOrUpdateAtSubscription: createAtSubscription,
      createOrUpdateAtManagementGroup: createAtManagementGroup,
      validateStackAtResourceGroup: validateAtResourceGroup,
      validateStackAtSubscription: validateAtSubscription,
      validateStackAtManagementGroup: validateAtManagementGroup,
      deleteAtResourceGroup,
      deleteAtSubscription,
      deleteAtManagementGroup,
    },
  })),
}));

const credential = {} as TokenCredential;
const stack = {
  properties: {
    outputs: { endpoint: { type: 'String', value: 'https://example.test' } },
    resources: [{ id: '/subscriptions/sub/providers/Test/widgets/one', type: 'Test/widgets' }],
  },
};
const validation = {
  properties: {
    correlationId: '00000000-0000-0000-0000-000000000001',
    validatedResources: stack.properties.resources,
  },
};
const deleteOptions = {
  unmanageActionManagementGroups: 'delete',
  unmanageActionResourceGroups: 'delete',
  unmanageActionResources: 'delete',
  unmanageActionResourcesWithoutDeleteSupport: 'fail',
};

function poller<T>(value: T): { pollUntilDone: jest.Mock<Promise<T>> } {
  return { pollUntilDone: jest.fn().mockResolvedValue(value) };
}

function createLiveSession(bicep: Partial<Bicep>): LiveBicepTestSession {
  const offline = Reflect.construct(BicepTestSession, [bicep]) as BicepTestSession;
  return Reflect.construct(LiveBicepTestSession, [offline, credential]) as LiveBicepTestSession;
}

function successfulBicep(parameters: Record<string, unknown> = {}): Partial<Bicep> {
  return {
    compileParams: jest.fn().mockResolvedValue({
      success: true,
      diagnostics: [],
      template: JSON.stringify({ resources: [] }),
      parameters: JSON.stringify({ parameters }),
    }),
  } as unknown as Partial<Bicep>;
}

describe('Live deployment helper', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    createAtResourceGroup.mockReturnValue(poller(stack));
    createAtSubscription.mockReturnValue(poller(stack));
    createAtManagementGroup.mockReturnValue(poller(stack));
    validateAtResourceGroup.mockReturnValue(poller(validation));
    validateAtSubscription.mockReturnValue(poller(validation));
    validateAtManagementGroup.mockReturnValue(poller(validation));
    deleteAtResourceGroup.mockReturnValue(poller(undefined));
    deleteAtSubscription.mockReturnValue(poller(undefined));
    deleteAtManagementGroup.mockReturnValue(poller(undefined));
  });

  it.each([
    {
      name: 'resource group',
      options: { filePath: 'main.bicepparam', subscriptionId: 'sub', resourceGroup: 'rg', location: 'westus' },
      expectedLocation: 'westus',
      expectedClientArguments: [credential, 'sub'],
      create: createAtResourceGroup,
      validate: validateAtResourceGroup,
      remove: deleteAtResourceGroup,
      leadingArguments: ['rg'],
    },
    {
      name: 'subscription',
      options: { filePath: 'main.bicepparam', subscriptionId: 'sub', location: 'eastus' },
      expectedLocation: 'eastus',
      expectedClientArguments: [credential, 'sub'],
      create: createAtSubscription,
      validate: validateAtSubscription,
      remove: deleteAtSubscription,
      leadingArguments: [],
    },
    {
      name: 'management group',
      options: { filePath: 'main.bicepparam', managementGroupId: 'mg', location: 'eastus' },
      expectedLocation: 'eastus',
      expectedClientArguments: [credential],
      create: createAtManagementGroup,
      validate: validateAtManagementGroup,
      remove: deleteAtManagementGroup,
      leadingArguments: ['mg'],
    },
  ])('creates, validates, and deletes at $name scope', async scenario => {
    const session = createLiveSession(successfulBicep());

    const validated = await session.validate(scenario.options as DeployOptions);
    const deployed = await session.deploy(scenario.options as DeployOptions);
    await deployed.teardown();

    expect(DeploymentStacksClient).toHaveBeenCalledWith(...scenario.expectedClientArguments);
    expect(validated).toEqual({
      isValid: true,
      resources: stack.properties.resources,
      correlationId: validation.properties.correlationId,
      error: undefined,
    });
    expect(deployed).toEqual(expect.objectContaining({
      succeeded: true,
      error: undefined,
      errorCode: undefined,
      errorMessage: undefined,
      outputs: { endpoint: 'https://example.test' },
      resources: stack.properties.resources,
    }));

    const validateArguments = scenario.validate.mock.calls[0];
    const createArguments = scenario.create.mock.calls[0];
    const stackName = createArguments[scenario.leadingArguments.length];
    const validationStackName = validateArguments[scenario.leadingArguments.length];
    expect(stackName).toMatch(/^bicep-test-[0-9a-f]{32}$/);
    expect(validationStackName).toMatch(/^bicep-test-[0-9a-f]{32}$/);
    expect(stackName).not.toBe(validationStackName);
    expect(createArguments.at(-1)).toEqual(expect.objectContaining({ location: scenario.expectedLocation }));
    expect(scenario.remove).toHaveBeenCalledWith(
      ...scenario.leadingArguments,
      stackName,
      deleteOptions,
    );
  });

  it('preserves compiled values and Key Vault references', async () => {
    const parameters = {
      environment: { value: 'test' },
      optionalValue: { value: null },
      secret: {
        reference: {
          keyVault: { id: '/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test' },
          secretName: 'password',
          secretVersion: 'version',
        },
      },
    };
    const session = createLiveSession(successfulBicep(parameters));

    await session.deploy({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
      resourceGroup: 'rg',
      stackName: 'stack',
      parameterOverrides: { environment: 'override' },
    });

    const compileParams = (session as unknown as {
      session: { bicep: { compileParams: jest.Mock } };
    }).session.bicep.compileParams;
    expect(compileParams).toHaveBeenCalledWith(expect.objectContaining({
      parameterOverrides: { environment: 'override' },
    }));
    expect(createAtResourceGroup.mock.calls[0][2].properties.parameters).toEqual(parameters);
  });

  it('returns rich validation errors', async () => {
    const error = {
      code: 'InvalidTemplate',
      message: 'The template is invalid.',
      target: 'resources[0]',
      details: [{ code: 'InvalidResource', message: 'The resource is invalid.' }],
    };
    validateAtResourceGroup.mockReturnValue(poller({ error, properties: {} }));
    const session = createLiveSession(successfulBicep());

    const result = await session.validate({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
      resourceGroup: 'rg',
    });

    expect(result).toEqual({
      isValid: false,
      resources: [],
      correlationId: undefined,
      error: { code: error.code, message: error.message, rawData: error },
    });
  });

  it('returns post-submission failures with cleanup available', async () => {
    const serviceBody = {
      error: {
        code: 'DeploymentStackOutOfSync',
        message: 'The stack is out of sync.',
        details: [{ code: 'ManagedResourceFailure', message: 'A managed resource failed.' }],
      },
    };
    createAtResourceGroup.mockReturnValue({
      pollUntilDone: jest.fn().mockRejectedValue(Object.assign(
        new Error(serviceBody.error.message),
        { code: serviceBody.error.code, response: { status: 409, bodyAsText: JSON.stringify(serviceBody) } },
      )),
    });
    const session = createLiveSession(successfulBicep());

    const result = await session.deploy({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
      resourceGroup: 'rg',
      stackName: 'failed-stack',
    });

    expect(result).toEqual(expect.objectContaining({
      succeeded: false,
      errorCode: serviceBody.error.code,
      errorMessage: serviceBody.error.message,
      outputs: {},
      resources: [],
      error: {
        code: serviceBody.error.code,
        message: serviceBody.error.message,
        rawData: serviceBody,
      },
    }));
    await Promise.all([result.teardown(), result.teardown()]);
    expect(deleteAtResourceGroup).toHaveBeenCalledTimes(1);
  });

  it('treats a 404 teardown response as success', async () => {
    deleteAtResourceGroup.mockReturnValue({
      pollUntilDone: jest.fn().mockRejectedValue({ statusCode: 404 }),
    });
    const session = createLiveSession(successfulBicep());
    const result = await session.deploy({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
      resourceGroup: 'rg',
    });

    await expect(result.teardown()).resolves.toBeUndefined();
  });

  it('rejects option and compilation errors before Azure submission', async () => {
    const compileParams = jest.fn().mockResolvedValue({
      success: false,
      diagnostics: [{ level: 'Error', code: 'BCP001', message: 'Invalid Bicep.' }],
    });
    const session = createLiveSession({ compileParams } as unknown as Bicep);

    await expect(session.deploy({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
    } as DeployOptions)).rejects.toThrow('location is required');
    await expect(session.deploy({
      filePath: 'main.bicepparam',
      subscriptionId: 'sub',
      resourceGroup: 'rg',
    })).rejects.toThrow('Error BCP001: Invalid Bicep.');
    expect(DeploymentStacksClient).not.toHaveBeenCalled();
  });

  it.each([
    {
      filePath: 'main.bicepparam',
      managementGroupId: 'mg',
      subscriptionId: 'sub',
      location: 'eastus',
    },
    {
      filePath: 'main.bicepparam',
      resourceGroup: 'rg',
    },
  ])('rejects an ambiguous or incomplete target before compilation', async invalidOptions => {
    const compileParams = jest.fn();
    const session = createLiveSession({ compileParams } as unknown as Bicep);

    await expect(session.deploy(invalidOptions as DeployOptions)).rejects.toThrow(TypeError);

    expect(compileParams).not.toHaveBeenCalled();
    expect(DeploymentStacksClient).not.toHaveBeenCalled();
  });

  it('forwards snapshots and disposes the owned offline session', async () => {
    const snapshot = jest.fn().mockResolvedValue({ predictedResources: [], diagnostics: [], outputs: {} });
    const dispose = jest.fn();
    const offline = { snapshot, dispose } as unknown as BicepTestSession;
    const session = Reflect.construct(
      LiveBicepTestSession,
      [offline, credential],
    ) as LiveBicepTestSession;

    await session.snapshot('main.bicepparam', 'tenant', 'sub', 'rg', 'eastus', 'deployment');
    session.dispose();

    expect(snapshot).toHaveBeenCalledWith('main.bicepparam', 'tenant', 'sub', 'rg', 'eastus', 'deployment');
    expect(dispose).toHaveBeenCalledTimes(1);
  });
});

File: packages/go/bicep-testing/deployment_test.go
package biceptesting

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"path/filepath"
	"regexp"
	"strings"
	"sync/atomic"
	"testing"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armdeploymentstacks"
	biceprpcclient "github.com/anthony-c-martin/bicep-testing/packages/go/bicep-rpc-client"
)

func TestLiveDeploySupportsAllScopes(t *testing.T) {
	tests := []struct {
		name                       string
		options                    DeployOptions
		expectedScope              deploymentScope
		expectedSubscription       string
		expectedResourceGroup      string
		expectedManagementGroup    string
		expectedLocation           string
		expectedClientSubscription string
	}{
		{
			name: "resource group",
			options: DeployOptions{
				FilePath:       "main.bicepparam",
				SubscriptionID: subscriptionID,
				ResourceGroup:  resourceGroup,
				Location:       "westus",
			},
			expectedScope:              deploymentScopeResourceGroup,
			expectedSubscription:       subscriptionID,
			expectedResourceGroup:      resourceGroup,
			expectedLocation:           "westus",
			expectedClientSubscription: subscriptionID,
		},
		{
			name: "subscription",
			options: DeployOptions{
				FilePath:       "main.bicepparam",
				SubscriptionID: subscriptionID,
				Location:       "eastus2",
			},
			expectedScope:              deploymentScopeSubscription,
			expectedSubscription:       subscriptionID,
			expectedLocation:           "eastus2",
			expectedClientSubscription: subscriptionID,
		},
		{
			name: "management group",
			options: DeployOptions{
				FilePath:          "main.bicepparam",
				ManagementGroupID: "mg-ops",
				Location:          "centralus",
			},
			expectedScope:              deploymentScopeManagementGroup,
			expectedManagementGroup:    "mg-ops",
			expectedLocation:           "centralus",
			expectedClientSubscription: "",
		},
	}

	for _, test := range tests {
		test := test
		t.Run(test.name, func(t *testing.T) {
			stackClient := &fakeDeploymentStackClient{}
			var actualClientSubscription string
			originalFactory := newDeploymentStackClient
			newDeploymentStackClient = func(actualSubscriptionID string, credential azcore.TokenCredential) (deploymentStackClient, error) {
				actualClientSubscription = actualSubscriptionID
				return stackClient, nil
			}
			t.Cleanup(func() { newDeploymentStackClient = originalFactory })

			bicep := &fakeBicepClient{compilation: successfulCompilation()}
			session := &LiveSession{session: &Session{client: bicep}, credential: fakeCredential{}}

			result, err := session.Deploy(context.Background(), test.options)
			if err != nil {
				t.Fatalf("Deploy returned an error: %v", err)
			}
			if result == nil || !result.Succeeded {
				t.Fatalf("Deploy result = %#v, want success", result)
			}
			if actualClientSubscription != test.expectedClientSubscription {
				t.Errorf("client subscription = %q, want %q", actualClientSubscription, test.expectedClientSubscription)
			}
			if stackClient.target.scope != test.expectedScope {
				t.Errorf("scope = %v, want %v", stackClient.target.scope, test.expectedScope)
			}
			if stackClient.target.subscriptionID != test.expectedSubscription {
				t.Errorf("subscription ID = %q, want %q", stackClient.target.subscriptionID, test.expectedSubscription)
			}
			if stackClient.target.resourceGroup != test.expectedResourceGroup {
				t.Errorf("resource group = %q, want %q", stackClient.target.resourceGroup, test.expectedResourceGroup)
			}
			if stackClient.target.managementGroupID != test.expectedManagementGroup {
				t.Errorf("management group = %q, want %q", stackClient.target.managementGroupID, test.expectedManagementGroup)
			}
			if stackClient.stack.Location == nil || *stackClient.stack.Location != test.expectedLocation {
				t.Errorf("stack location = %v, want %q", stackClient.stack.Location, test.expectedLocation)
			}
			if !regexp.MustCompile(`^bicep-test-[0-9a-f]{32}$`).MatchString(stackClient.stackName) {
				t.Errorf("default stack name = %q, want bicep-test-<32 hex chars>", stackClient.stackName)
			}
		})
	}
}

func TestLiveDeployCompilesAndPreservesParameters(t *testing.T) {
	stackClient := &fakeDeploymentStackClient{}
	originalFactory := newDeploymentStackClient
	newDeploymentStackClient = func(actualSubscriptionID string, credential azcore.TokenCredential) (deploymentStackClient, error) {
		if actualSubscriptionID != subscriptionID {
			t.Errorf("subscription ID = %q, want %q", actualSubscriptionID, subscriptionID)
		}
		return stackClient, nil
	}
	t.Cleanup(func() { newDeploymentStackClient = originalFactory })

	bicep := &fakeBicepClient{compilation: biceprpcclient.CompileParamsResponse{
		Success:    true,
		Template:   `{"resources":[]}`,
		Parameters: `{"parameters":{"message":{"value":"hello"},"optionalValue":{"value":null},"secret":{"reference":{"keyVault":{"id":"/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test"},"secretName":"password","secretVersion":"v1"}}}}`,
	}}
	session := &LiveSession{session: &Session{client: bicep}, credential: fakeCredential{}}
	overrides := map[string]json.RawMessage{"message": json.RawMessage(`"override"`)}
	result, err := session.Deploy(context.Background(), DeployOptions{
		FilePath:           "main.bicepparam",
		SubscriptionID:     subscriptionID,
		ResourceGroup:      resourceGroup,
		StackName:          "test-stack-parameters",
		ParameterOverrides: overrides,
	})
	if err != nil {
		t.Fatalf("Deploy returned an error: %v", err)
	}
	if !filepath.IsAbs(bicep.request.Path) {
		t.Errorf("compile path = %q, want an absolute path", bicep.request.Path)
	}
	if string(bicep.request.ParameterOverrides["message"]) != `"override"` {
		t.Errorf("parameter override = %s, want override", bicep.request.ParameterOverrides["message"])
	}
	if stackClient.target.resourceGroup != resourceGroup || stackClient.stackName != "test-stack-parameters" {
		t.Errorf("stack target = %s/%s, want %s/test-stack-parameters", stackClient.target.resourceGroup, stackClient.stackName, resourceGroup)
	}
	if actual := stackClient.stack.Properties.Parameters["message"].Value; actual != "hello" {
		t.Errorf("deployment parameter = %v, want hello", actual)
	}
	if actual := stackClient.stack.Properties.Parameters["optionalValue"].Value; !azcore.IsNullValue(actual) {
		t.Errorf("optional deployment parameter = %#v, want explicit null", actual)
	}
	encodedNull, err := json.Marshal(stackClient.stack.Properties.Parameters["optionalValue"])
	if err != nil {
		t.Fatalf("marshal null deployment parameter: %v", err)
	}
	if string(encodedNull) != `{"value":null}` {
		t.Errorf("null deployment parameter = %s, want explicit null value", encodedNull)
	}
	secret := stackClient.stack.Properties.Parameters["secret"].Reference
	if secret == nil || secret.KeyVault == nil || *secret.KeyVault.ID != "/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test" || *secret.SecretName != "password" || *secret.SecretVersion != "v1" {
		t.Errorf("secret parameter = %#v, want a Key Vault reference", secret)
	}
	if actual := result.Outputs["endpoint"]; actual != "https://example.test" {
		t.Errorf("endpoint output = %v, want https://example.test", actual)
	}
	if len(result.Resources) != 1 || result.Resources[0].Type != "Microsoft.Storage/storageAccounts" {
		t.Errorf("resources = %#v, want one storage account", result.Resources)
	}
}

func TestValidateReturnsRichErrorResult(t *testing.T) {
	stackClient := &fakeDeploymentStackClient{
		validation: armdeploymentstacks.DeploymentStackValidateResult{
			Error: &armdeploymentstacks.ErrorDetail{
				Code:    stringPointer("InvalidTemplate"),
				Message: stringPointer("The template is invalid."),
			},
			Properties: &armdeploymentstacks.DeploymentStackValidateProperties{
				CorrelationID: stringPointer("00000000-0000-0000-0000-000000000001"),
			},
		},
	}
	originalFactory := newDeploymentStackClient
	newDeploymentStackClient = func(string, azcore.TokenCredential) (deploymentStackClient, error) {
		return stackClient, nil
	}
	t.Cleanup(func() { newDeploymentStackClient = originalFactory })

	session := &LiveSession{session: &Session{client: &fakeBicepClient{compilation: successfulCompilation()}}, credential: fakeCredential{}}

	result, err := session.Validate(context.Background(), DeployOptions{
		FilePath:       "main.bicepparam",
		SubscriptionID: subscriptionID,
		ResourceGroup:  resourceGroup,
		StackName:      "validate-stack",
	})
	if err != nil {
		t.Fatalf("Validate returned an error: %v", err)
	}
	if result.IsValid {
		t.Fatalf("IsValid = true, want false")
	}
	if result.CorrelationID != "00000000-0000-0000-0000-000000000001" {
		t.Errorf("correlation ID = %q", result.CorrelationID)
	}
	if result.Error == nil || result.Error.Code != "InvalidTemplate" || result.Error.Message != "The template is invalid." {
		t.Fatalf("error = %#v, want code/message from validation", result.Error)
	}
	if !strings.Contains(string(result.Error.RawData), "InvalidTemplate") {
		t.Errorf("raw data = %s, want serialized service error", result.Error.RawData)
	}
}

func TestDeployReturnsFailedResultAfterSubmissionAndKeepsCleanup(t *testing.T) {
	body := `{"error":{"code":"DeploymentStackOutOfSync","message":"The stack is out of sync.","details":[{"code":"ManagedResourceFailure","message":"A managed resource failed."}]}}`
	stackClient := &fakeDeploymentStackClient{
		createError: &azcore.ResponseError{
			ErrorCode:  "DeploymentStackOutOfSync",
			StatusCode: http.StatusConflict,
			RawResponse: &http.Response{
				StatusCode: http.StatusConflict,
				Body:       io.NopCloser(strings.NewReader(body)),
			},
		},
	}
	originalFactory := newDeploymentStackClient
	newDeploymentStackClient = func(string, azcore.TokenCredential) (deploymentStackClient, error) {
		return stackClient, nil
	}
	t.Cleanup(func() { newDeploymentStackClient = originalFactory })

	session := &LiveSession{session: &Session{client: &fakeBicepClient{compilation: successfulCompilation()}}, credential: fakeCredential{}}
	result, err := session.Deploy(context.Background(), DeployOptions{
		FilePath:       "main.bicepparam",
		SubscriptionID: subscriptionID,
		ResourceGroup:  resourceGroup,
		StackName:      "failed-stack",
	})
	if err != nil {
		t.Fatalf("Deploy returned an error: %v", err)
	}
	if result.Succeeded {
		t.Fatalf("Succeeded = true, want false")
	}
	if result.ErrorCode != "DeploymentStackOutOfSync" || result.ErrorMessage != "The stack is out of sync." {
		t.Fatalf("error fields = %#v", result)
	}
	if result.Error == nil || !strings.Contains(string(result.Error.RawData), "ManagedResourceFailure") {
		t.Fatalf("error = %#v, want raw service body", result.Error)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("Teardown returned an error: %v", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("second Teardown returned an error: %v", err)
	}
	if stackClient.deleteCalls != 1 {
		t.Errorf("delete calls = %d, want 1", stackClient.deleteCalls)
	}
}

func TestDeployRejectsBadOptionsAndPresubmissionFailures(t *testing.T) {
	t.Run("bad options", func(t *testing.T) {
		bicep := &fakeBicepClient{compilation: successfulCompilation()}
		session := &LiveSession{session: &Session{client: bicep}, credential: fakeCredential{}}

		_, err := session.Deploy(context.Background(), DeployOptions{FilePath: "main.bicepparam", SubscriptionID: subscriptionID})
		if err == nil || !strings.Contains(err.Error(), "location is required") {
			t.Fatalf("error = %v, want location validation", err)
		}
		if bicep.compileCalls != 0 {
			t.Errorf("compile calls = %d, want 0", bicep.compileCalls)
		}
	})

	t.Run("compile failure", func(t *testing.T) {
		bicep := &fakeBicepClient{compilation: biceprpcclient.CompileParamsResponse{Success: false, Diagnostics: []map[string]any{{"level": "Error", "code": "BCP001", "message": "Invalid Bicep."}}}}
		session := &LiveSession{session: &Session{client: bicep}, credential: fakeCredential{}}

		_, err := session.Deploy(context.Background(), DeployOptions{FilePath: "main.bicepparam", SubscriptionID: subscriptionID, ResourceGroup: resourceGroup, StackName: "x"})
		if err == nil || !strings.Contains(err.Error(), "BCP001") {
			t.Fatalf("error = %v, want compile diagnostics", err)
		}
	})

	t.Run("client construction failure", func(t *testing.T) {
		originalFactory := newDeploymentStackClient
		newDeploymentStackClient = func(string, azcore.TokenCredential) (deploymentStackClient, error) {
			return nil, errors.New("factory failure")
		}
		t.Cleanup(func() { newDeploymentStackClient = originalFactory })

		session := &LiveSession{session: &Session{client: &fakeBicepClient{compilation: successfulCompilation()}}, credential: fakeCredential{}}
		_, err := session.Deploy(context.Background(), DeployOptions{FilePath: "main.bicepparam", SubscriptionID: subscriptionID, ResourceGroup: resourceGroup, StackName: "x"})
		if err == nil || !strings.Contains(err.Error(), "factory failure") {
			t.Fatalf("error = %v, want factory failure", err)
		}
	})
}

func TestLiveSessionForwardsSnapshotAndClose(t *testing.T) {
	t.Parallel()

	bicep := &fakeBicepClient{snapshot: biceprpcclient.GetSnapshotResponse{Snapshot: `{"predictedResources":[],"diagnostics":[],"outputs":{}}`}}
	live := &LiveSession{session: &Session{client: bicep}, credential: fakeCredential{}}

	_, err := live.Snapshot(context.Background(), "main.bicepparam", SnapshotMetadata{TenantID: tenantID})
	if err != nil {
		t.Fatalf("Snapshot returned an error: %v", err)
	}
	if err := live.Close(); err != nil {
		t.Fatalf("Close returned an error: %v", err)
	}
	if bicep.closed != 1 {
		t.Errorf("close calls = %d, want 1", bicep.closed)
	}
}

func TestTeardownRetriesAfterFailure(t *testing.T) {
	t.Parallel()

	stackClient := &fakeDeploymentStackClient{deleteErrors: []error{context.Canceled, nil}}
	result := &DeployResult{client: stackClient, target: deploymentTarget{scope: deploymentScopeResourceGroup, resourceGroup: resourceGroup}, stackName: "test-stack"}

	if err := result.Teardown(context.Background()); !errors.Is(err, context.Canceled) {
		t.Fatalf("first Teardown error = %v, want context canceled", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("second Teardown returned an error: %v", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("third Teardown returned an error: %v", err)
	}
	if stackClient.deleteCalls != 2 {
		t.Errorf("delete calls = %d, want 2", stackClient.deleteCalls)
	}
}

func TestTeardownSharesActiveDeletionAndAllowsWaiterCancellation(t *testing.T) {
	t.Parallel()

	stackClient := &blockingDeploymentStackClient{
		started: make(chan struct{}),
		release: make(chan struct{}),
	}
	result := &DeployResult{client: stackClient, target: deploymentTarget{scope: deploymentScopeResourceGroup, resourceGroup: resourceGroup}, stackName: "test-stack"}
	firstDone := make(chan error, 1)
	go func() {
		firstDone <- result.Teardown(context.Background())
	}()
	<-stackClient.started

	waiterCtx, cancelWaiter := context.WithCancel(context.Background())
	waiterDone := make(chan error, 1)
	go func() {
		waiterDone <- result.Teardown(waiterCtx)
	}()
	cancelWaiter()
	if err := <-waiterDone; !errors.Is(err, context.Canceled) {
		t.Fatalf("waiting Teardown error = %v, want context canceled", err)
	}
	if calls := stackClient.deleteCalls.Load(); calls != 1 {
		t.Fatalf("delete calls while active = %d, want 1", calls)
	}

	close(stackClient.release)
	if err := <-firstDone; err != nil {
		t.Fatalf("active Teardown returned an error: %v", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("completed Teardown returned an error: %v", err)
	}
	if calls := stackClient.deleteCalls.Load(); calls != 1 {
		t.Errorf("delete calls = %d, want 1", calls)
	}
}

func TestTeardownTreatsNotFoundAsSuccess(t *testing.T) {
	t.Parallel()

	stackClient := &fakeDeploymentStackClient{deleteErrors: []error{&azcore.ResponseError{StatusCode: http.StatusNotFound}}}
	result := &DeployResult{client: stackClient, target: deploymentTarget{scope: deploymentScopeResourceGroup, resourceGroup: resourceGroup}, stackName: "missing-stack"}

	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("Teardown returned an error: %v", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("second Teardown returned an error: %v", err)
	}
	if stackClient.deleteCalls != 1 {
		t.Errorf("delete calls = %d, want 1", stackClient.deleteCalls)
	}
}

func TestNewLiveSessionRejectsNilCredential(t *testing.T) {
	t.Parallel()

	_, err := NewLiveSession(context.Background(), "0.46.1", nil)
	if err == nil || !strings.Contains(err.Error(), "credential") {
		t.Fatalf("error = %v, want credential validation", err)
	}
}

type fakeCredential struct{}

func (fakeCredential) GetToken(context.Context, policy.TokenRequestOptions) (azcore.AccessToken, error) {
	return azcore.AccessToken{Token: "token", ExpiresOn: time.Now().Add(time.Hour)}, nil
}

type fakeBicepClient struct {
	request      biceprpcclient.CompileParamsRequest
	compilation  biceprpcclient.CompileParamsResponse
	compileCalls int
	snapshot     biceprpcclient.GetSnapshotResponse
	closed       int
}

func (client *fakeBicepClient) CompileParams(_ context.Context, request biceprpcclient.CompileParamsRequest) (biceprpcclient.CompileParamsResponse, error) {
	client.request = request
	client.compileCalls++
	return client.compilation, nil
}

func (client *fakeBicepClient) GetSnapshot(context.Context, biceprpcclient.GetSnapshotRequest) (biceprpcclient.GetSnapshotResponse, error) {
	return client.snapshot, nil
}

func (client *fakeBicepClient) Close() error {
	client.closed++
	return nil
}

type fakeDeploymentStackClient struct {
	target       deploymentTarget
	stackName    string
	stack        armdeploymentstacks.DeploymentStack
	validation   armdeploymentstacks.DeploymentStackValidateResult
	createError  error
	deleteCalls  int
	deleteErrors []error
}

func (client *fakeDeploymentStackClient) createOrUpdate(_ context.Context, target deploymentTarget, stackName string, stack armdeploymentstacks.DeploymentStack) (armdeploymentstacks.DeploymentStack, error) {
	client.target = target
	client.stackName = stackName
	client.stack = stack
	if client.createError != nil {
		return armdeploymentstacks.DeploymentStack{}, client.createError
	}
	resourceID := "/subscriptions/sub/resourceGroups/rg/providers/Microsoft.Storage/storageAccounts/example"
	return armdeploymentstacks.DeploymentStack{Properties: &armdeploymentstacks.DeploymentStackProperties{
		Outputs: map[string]any{"endpoint": map[string]any{"value": "https://example.test"}},
		Resources: []*armdeploymentstacks.ManagedResourceReference{
			{ID: &resourceID},
		},
	}}, nil
}

func (client *fakeDeploymentStackClient) validate(_ context.Context, target deploymentTarget, stackName string, stack armdeploymentstacks.DeploymentStack) (armdeploymentstacks.DeploymentStackValidateResult, error) {
	client.target = target
	client.stackName = stackName
	client.stack = stack
	if client.validation.Properties == nil {
		client.validation.Properties = &armdeploymentstacks.DeploymentStackValidateProperties{}
	}
	return client.validation, nil
}

func (client *fakeDeploymentStackClient) delete(_ context.Context, target deploymentTarget, stackName string) error {
	client.target = target
	client.stackName = stackName
	client.deleteCalls++
	if len(client.deleteErrors) > 0 {
		err := client.deleteErrors[0]
		client.deleteErrors = client.deleteErrors[1:]
		return err
	}
	return nil
}

type blockingDeploymentStackClient struct {
	started     chan struct{}
	release     chan struct{}
	deleteCalls atomic.Int32
}

func (*blockingDeploymentStackClient) createOrUpdate(context.Context, deploymentTarget, string, armdeploymentstacks.DeploymentStack) (armdeploymentstacks.DeploymentStack, error) {
	return armdeploymentstacks.DeploymentStack{}, nil
}

func (*blockingDeploymentStackClient) validate(context.Context, deploymentTarget, string, armdeploymentstacks.DeploymentStack) (armdeploymentstacks.DeploymentStackValidateResult, error) {
	return armdeploymentstacks.DeploymentStackValidateResult{}, nil
}

func (client *blockingDeploymentStackClient) delete(ctx context.Context, _ deploymentTarget, _ string) error {
	client.deleteCalls.Add(1)
	close(client.started)
	select {
	case <-client.release:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

func successfulCompilation() biceprpcclient.CompileParamsResponse {
	return biceprpcclient.CompileParamsResponse{
		Success:    true,
		Template:   `{"resources":[]}`,
		Parameters: `{"parameters":{"message":{"value":"hello"}}}`,
	}
}

func stringPointer(value string) *string {
	return &value
}
